File: cc/026.cc
// Project Euler Problem 026
//
// Find the value of d < 1000 for which 1/d contains the longest
// recurring cycle.

#include <cstdio>

int repeat_length(int i) {
  int repeats[1000] = {0};
  int index = 0;
  int q = 1;
  while (true) {
    while (q <= i) {
      index++;
      q *= 10;
    }
    if ((q % i) == 0)
      return 0;
    q -= i * (q/i);
    if (repeats[q])
      return index - repeats[q];
    repeats[q] = index;
  }
}

int main(int argc, const char* argv[]) {
  int max = 0;
  int maxlen = 0;
  for (int i=2; i<1000; i++) {
    int len = repeat_length(i);
    if (len > maxlen) {
      maxlen = len;
      max = i;
    }
  }
  printf("%d\n", max);
}

File: go/problems/026.go
// Project Euler Problem 026
//
// Find the value of d < 1000 for which 1/d contains the longest
// recurring cycle.

package problems

import "fmt"

func repeatLength(i int) int {
	positionForQuotient := make(map[int]int)
	index, q := 0, 1
	for {
		for ; q < i ; {
			index++
			q *= 10
		}
		// Non-repeating
		if ((q % i) == 0) {
			return 0
		}
		q -= i * (q/i)
		if positionForQuotient[q] > 0 {
			return index - positionForQuotient[q]
		}
		positionForQuotient[q] = index
	}
	return 0
}

func Problem026() string {
	max, maxlen := 0, 0
	for i := 7; i < 1000; i++ {
		if length := repeatLength(i); length > maxlen {
			maxlen, max = length, i
		}
	}
	return fmt.Sprintf("%d", max)
}

func init() {
     Functions[26] = Problem026
}
